src/machma/api.py

```diff
@@ -0,0 +1,184 @@
+        .group_by(Exercise.exercise_name)
+        .outerjoin(UserReps)
+    )
+
+
+def _get_user_reps(user_id: int):
+    # Ensure that the user actually exists.
+    user = session.query(User).filter(User.user_id == user_id).first()
+    if not user:
+        raise UserDoesNotExistError(user_id)
+    # Get a subquery for the reps of that user.
+    user_reps = user.reps.subquery()
+    # Left outer join the reps on th exercises.
+    query = (session
+        .query(Exercise.exercise_name, F.coalesce(user_reps.c.reps, 0).label('reps'))
+        .select_from(Exercise)
+        .outerjoin(user_reps)
+    )
+    return query
+
+
+def _get_user_todo_reps(user_id: int):
+    max_reps = _get_max_reps().subquery()
+    user_reps = _get_user_reps(user_id).subquery()
+    return (
+        session
+        .query(
+            max_reps.c.exercise_name.label('exercise_name'),
+            (max_reps.c.reps - user_reps.c.reps).label('reps'),
+        )
+        .select_from(
+            max_reps.join(user_reps, max_reps.c.exercise_name == user_reps.c.exercise_name)
+        )
+    )
+
+
+def _get_reps_for_exercise(query, exercise: str) -> int:
+    # TODO(NiklasRosenstein): This doesn't seem quite like the right way to do this.
+    #   Check https://stackoverflow.com/q/63728782/791713 for suggestions.
+    query = query.subquery()
+    query = session.query(*query.c).filter(query.c.exercise_name == exercise)
+    try:
+        return query.one()[1]
+    except NoResultFound:
+        raise ExerciseDoesNotExistError(exercise)
+
+
+def get_user(user_id: int) -> Optional[Dict[str, Any]]:
+    user = session.query(User).filter(User.user_id == user_id).first()
+    if user:
+        return {
+            'id': user.user_id,
+            'user_name': user.user_name,
+            'first_name': user.first_name,
+            'last_name': user.last_name}
+    return None
+
+
+def has_user(user_id: int) -> bool:
+    return session.query(User).filter(User.user_id == user_id).count() != 0
+
+
+def get_max_reps() -> Dict[str, int]:
+    return dict(_get_max_reps())
+
+
+def get_max_reps_for_exercise(exercise: str) -> int:
+    return _get_reps_for_exercise(_get_max_reps(), exercise)
+
+
+def get_user_reps(user_id: int) -> Dict[str, int]:
+    return dict(_get_user_reps(user_id))
+
+
+def get_user_reps_for_exercise(user_id: int, exercise: str):
+    return _get_reps_for_exercise(_get_user_reps(user_id), exercise)
+
+
+def get_user_todo_reps(user_id: int) -> Dict[str, int]:
+    return dict(_get_user_todo_reps(user_id))
+
+
+def get_user_todo_reps_for_exercise(user_id: int, exercise: str) -> int:
+    return _get_reps_for_exercise(_get_user_todo_reps(user_id), exercise)
+
+
+def add_to_user_reps(user_id: int, exercise: str, reps: int) -> None:
+    if not has_exercise(exercise):
+        raise ExerciseDoesNotExistError(exercise)
+    (db.get(UserReps, user_id=user_id, exercise_name=exercise)
```

Comment:
Unnecessary parentheses?

Reply:
They're necessary unless you want to put everything on a single line or add trailing backslashes. Personally I prefer the parentheses because the backslashes look ugly and make editing the code harder.